File: core/src/main/java/org/jdbi/v3/core/extension/Extensions.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jdbi.v3.core.extension;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Function;

import org.jdbi.v3.core.config.ConfigRegistry;
import org.jdbi.v3.core.config.JdbiConfig;
import org.jdbi.v3.core.extension.annotation.UseExtensionHandler;
import org.jdbi.v3.core.extension.annotation.UseExtensionHandlerCustomizer;
import org.jdbi.v3.core.internal.CopyOnWriteHashMap;
import org.jdbi.v3.meta.Alpha;
import org.jdbi.v3.meta.Beta;

import static org.jdbi.v3.core.extension.ExtensionFactory.FactoryFlag.NON_VIRTUAL_FACTORY;

/**
 * Configuration class for defining {@code Jdbi} extensions via {@link ExtensionFactory}
 * instances.
 */
public class Extensions implements JdbiConfig<Extensions> {

    private final List<ExtensionFactoryDelegate> extensionFactories;

    // Shared with the config this instance was copied from, so metadata computed inside a forked config (e.g.
    // a per-request Handle via handle.attach()) is reused by every other config forked from the same root.
    // Deep-copying it, as every other field is, left an attach()-only workload recomputing metadata on every
    // fresh Handle (jdbi/jdbi#2991). Metadata depends on the registrations here and on other config types read
    // through the registry while building it (e.g. Handlers, HandlerDecorators in sqlobject); any change to
    // such an input must invalidateMetadataCache() first so stale metadata is not served afterwards.
    // Annotation-driven config (e.g. @RegisterRowMapper) is exempt: it is a config customizer replayed per
    // instance at attach time, not baked into cached metadata.
    private volatile Map<Class<?>, ExtensionMetadata> extensionMetadataCache;

    private final List<ExtensionHandlerCustomizer> extensionHandlerCustomizers;
    private final List<ExtensionHandlerFactory> extensionHandlerFactories;
    private final List<ConfigCustomizerFactory> configCustomizerFactories;

    // Read while computing metadata (see ExtensionMetadata.isFailFast) on handle threads other than the one
    // that configured this instance, now that findMetadata runs against the shared cache. AtomicBoolean gives
    // the needed safe publication (a plain field trips SpotBugs AT_STALE_THREAD_WRITE_OF_PRIMITIVE). They are
    // final so the reference publishes safely via the copy constructor.
    private final AtomicBoolean allowProxy = new AtomicBoolean();
    private final AtomicBoolean failFast = new AtomicBoolean();

    private ConfigRegistry registry;

    /**
     * Creates a new instance.
     * <ul>
     * <li>registers extension handlers factories for bridge and interface default methods and for the {@link UseExtensionHandler} annotation.</li>
     * <li>registers extension handler customizers for the {@link UseExtensionHandlerCustomizer} annotation.</li>
     * <li>registers extension configurer factories for {@link org.jdbi.v3.core.extension.annotation.UseExtensionConfigurer} annotation.</li>
     * </ul>
     */
    public Extensions() {
        extensionFactories = new CopyOnWriteArrayList<>();
        extensionMetadataCache = new CopyOnWriteHashMap<>();
        extensionHandlerCustomizers = new CopyOnWriteArrayList<>();
        extensionHandlerFactories = new CopyOnWriteArrayList<>();
        configCustomizerFactories = new CopyOnWriteArrayList<>();

        allowProxy.set(true);
        failFast.set(false);

        // default handler factories for bridge and default methods
        internalRegisterHandlerFactory(DefaultMethodExtensionHandlerFactory.INSTANCE);
        internalRegisterHandlerFactory(BridgeMethodExtensionHandlerFactory.INSTANCE);

        // default handler factory for the UseExtensionHandler annotation.
        registerHandlerFactory(UseAnnotationExtensionHandlerFactory.INSTANCE);

        // default handler customizer for the UseExtensionHandlerCustomizer annotation.
        registerHandlerCustomizer(UseAnnotationExtensionHandlerCustomizer.INSTANCE);

        registerConfigCustomizerFactory(UseAnnotationConfigCustomizerFactory.INSTANCE);
    }

    /**
     * Create an extension configuration by cloning another.
     *
     * @param that the configuration to clone
     */
    private Extensions(Extensions that) {
        extensionFactories = new CopyOnWriteArrayList<>(that.extensionFactories);
        // Share the parent's cache reference rather than copying it. See the field comment.
        extensionMetadataCache = that.extensionMetadataCache;
        extensionHandlerCustomizers = new CopyOnWriteArrayList<>(that.extensionHandlerCustomizers);
        extensionHandlerFactories = new CopyOnWriteArrayList<>(that.extensionHandlerFactories);
        configCustomizerFactories = new CopyOnWriteArrayList<>(that.configCustomizerFactories);

        allowProxy.set(that.allowProxy.get());
        failFast.set(that.failFast.get());
    }

    @Override
    public void setRegistry(ConfigRegistry registry) {
        this.registry = registry;
    }

    /**
     * Register a {@link ExtensionFactory} instance with the extension framework.
     *
     * @param factory the factory to register
     * @return This instance
     */
    public Extensions register(ExtensionFactory factory) {
        invalidateMetadataCache();
        extensionFactories.add(0, new ExtensionFactoryDelegate(factory));
        return this;
    }

    /**
     * Registers a global {@link ExtensionHandlerFactory} instance. This factory is registered globally and will be used
     * with all registered {@link ExtensionFactory} instances.
     * @param extensionHandlerFactory The {@link ExtensionHandlerFactory} to register
     * @return This instance
     *
     * @since 3.38.0
     */
    @Alpha
    public Extensions registerHandlerFactory(ExtensionHandlerFactory extensionHandlerFactory) {
        invalidateMetadataCache();
        return internalRegisterHandlerFactory(FilteringExtensionHandlerFactory.forDelegate(extensionHandlerFactory));
    }

    /**
     * Registers a global {@link ExtensionHandlerCustomizer} instance. This customizer is registered globally and will be used
     * with all registered {@link ExtensionFactory} instances.
     * @param extensionHandlerCustomizer The {@link ExtensionHandlerCustomizer} to register
     * @return This instance
     *
     * @since 3.38.0
     */
    @Alpha
    public Extensions registerHandlerCustomizer(ExtensionHandlerCustomizer extensionHandlerCustomizer) {
        invalidateMetadataCache();
        extensionHandlerCustomizers.add(0, extensionHandlerCustomizer);
        return this;
    }

    /**
     * Registers a global {@link ConfigCustomizerFactory} instance. This factory is registered globally and will be used
     * with all registered {@link ExtensionFactory} instances.
     * @param configCustomizerFactory The {@link ConfigCustomizerFactory} to register
     * @return This instance
     *
     * @since 3.38.0
     */
    @Alpha
    public Extensions registerConfigCustomizerFactory(ConfigCustomizerFactory configCustomizerFactory) {
        invalidateMetadataCache();
        configCustomizerFactories.add(0, configCustomizerFactory);
        return this;
    }

    /**
     * Invalidates this configuration's extension metadata cache. Call this before mutating any state that
     * {@linkplain #findMetadata computed metadata} depends on, so that metadata computed under the previous
     * configuration is not served afterwards.
     * <p>
     * The cache is shared across copies of this configuration (see {@link #findMetadata}). Copies made before
     * this call keep their existing cache — a valid snapshot as of their creation — so a handle still sees
     * configuration as of the moment it was opened; this configuration and later copies recompute.
     * <p>
     * The {@code register*} methods here call this for their own registrations. A configuration type in
     * another module whose contents feed metadata computation must call it from its own mutators, via
     * {@code registry.get(Extensions.class).invalidateMetadataCache()} (the sqlobject {@code Handlers} and
     * {@code HandlerDecorators} do so).
     *
     * @since 3.55.0
     */
    @Alpha
    public void invalidateMetadataCache() {
        // Replace, don't clear: copies share this map reference and must keep their snapshot, so it must not
        // be mutated in place.
        extensionMetadataCache = new CopyOnWriteHashMap<>();
    }

    /**
     * Returns true if an extension is registered for the given extension type.
     *
     * @param extensionType the type to query. Must not be null
     * @return true if a registered extension factory handles the type
     */
    public boolean hasExtensionFor(Class<?> extensionType) {
        return findFactoryFor(extensionType).isPresent();
    }

    /**
     * Create an extension instance if a factory accepts the extension type.
     * <br>
     * <b>This method requires access to a {@link HandleSupplier}, which is only useful either from
     * within an extension implementation of inside the Jdbi code. It should rarely be called by
     * user code.</b>
     *
     * @param <E>            the extension type to create
     * @param extensionType  the extension type to create
     * @param handleSupplier A handle supplier object
     * @return an attached extension instance if a factory is found, {@link Optional#empty()} otherwise
     */
    public <E> Optional<E> findFor(Class<E> extensionType, HandleSupplier handleSupplier) {
        return findFactoryFor(extensionType)
                .map(factory -> factory.attach(extensionType, handleSupplier));
    }

    private Optional<ExtensionFactory> findFactoryFor(Class<?> extensionType) {
        for (ExtensionFactory factory : extensionFactories) {
            if (factory.accepts(extensionType)) {
                return Optional.of(factory);
            }
        }

        return Optional.empty();
    }

    /**
     * Find the registered factory of the given type. The factory returned from this call
     * may not be the same instance that was registered with {@link Extensions#register(ExtensionFactory)}.
     *
     * @param factoryType the factory's type to find
     * @return the found factory, if any or {@link Optional#empty()} otherwise
     */
    public Optional<ExtensionFactory> findFactory(Class<? extends ExtensionFactory> factoryType) {
        for (ExtensionFactoryDelegate factory : extensionFactories) {
            if (factoryType.isInstance(factory.getDelegatedFactory())) {
                return Optional.of(factory);
            }
        }

        return Optional.empty();
    }

    /**
     * Retrieves all extension metadata for a specific extension type.
     *
     * @param extensionType    The extension type
     * @param extensionFactory The extension factory for this extension type
     * @return A {@link ExtensionMetadata} object describing the extension handlers and customizers for this extension type
     *
     * @since 3.38.0
     */
    @Alpha
    public ExtensionMetadata findMetadata(Class<?> extensionType, ExtensionFactory extensionFactory) {
        // Racing first-computations may build metadata more than once (computeIfAbsent runs the mapping
        // function under a retry loop), which is harmless because metadata is a pure function of the config.
        return extensionMetadataCache.computeIfAbsent(extensionType, createMetadata(extensionFactory));
    }

    private Extensions internalRegisterHandlerFactory(ExtensionHandlerFactory extensionHandlerFactory) {
        extensionHandlerFactories.add(0, extensionHandlerFactory);
        return this;
    }

    private Function<Class<?>, ExtensionMetadata> createMetadata(ExtensionFactory extensionFactory) {
        return extensionType -> {

            ExtensionMetadata.Builder builder = ExtensionMetadata.builder(extensionType);

            // prep the extension handler set for this factory
            extensionFactory.getExtensionHandlerFactories(registry).forEach(builder::addExtensionHandlerFactory);
            extensionHandlerFactories.forEach(builder::addExtensionHandlerFactory);

            // InstanceExtensionHandlerFactory for non-virtual factories. These have a backing object and can invoke methods on those objects.
            if (extensionFactory.getFactoryFlags().contains(NON_VIRTUAL_FACTORY)) {
                builder.addExtensionHandlerFactory(InstanceExtensionHandlerFactory.INSTANCE);
            }

            // prep the extension customizer set for this factory
            extensionFactory.getExtensionHandlerCustomizers(registry).forEach(builder::addExtensionHandlerCustomizer);
            extensionHandlerCustomizers.forEach(builder::addExtensionHandlerCustomizer);

            // prep the extension configurer set for this factory
            extensionFactory.getConfigCustomizerFactories(registry).forEach(builder::addConfigCustomizerFactory);
            configCustomizerFactories.forEach(builder::addConfigCustomizerFactory);

            // build metadata
            extensionFactory.buildExtensionMetadata(builder);

            return builder.build();
        };
    }

    /**
     * Allow using {@link java.lang.reflect.Proxy} to implement extensions.
     *
     * @param allowProxy whether to allow use of Proxy types
     * @return this
     */
    @Beta
    public Extensions setAllowProxy(boolean allowProxy) {
        this.allowProxy.set(allowProxy);
        return this;
    }

    /**
     * Returns whether Proxy classes are allowed to be used.
     *
     * @return whether Proxy classes are allowed to be used.
     */
    @Beta
    public boolean isAllowProxy() {
        return allowProxy.get();
    }

    /**
     * Fail fast if any method in an Extension object is misconfigured and can not be warmed. Default is to
     * fail when a method is used for the first time.
     *
     * @return this
     * @since 3.39.0
     */
    @Alpha
    public Extensions failFast() {
        this.failFast.set(true);

        return this;
    }

    /**
     * Returns true if misconfigured extension objects fail fast.
     *
     * @return True if misconfigured extension objects fail fast.
     * @since 3.39.0
     */
    @Alpha
    public boolean isFailFast() {
        return failFast.get();
    }


    @Override
    public Extensions createCopy() {
        return new Extensions(this);
    }

    /**
     * Throw if proxy creation is disallowed.
     */
    @Beta
    public void onCreateProxy() {
        if (!isAllowProxy()) {
            throw new IllegalStateException(
                    "Creating onDemand proxy disallowed. Ensure @GenerateSqlObject annotation is being processed by `jdbi3-generator` annotation processor.");
        }
    }
}

File: kotlin/src/main/kotlin/org/jdbi/v3/core/kotlin/Extensions.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
@file:Suppress("TooManyFunctions")

package org.jdbi.v3.core.kotlin

import org.jdbi.v3.core.Handle
import org.jdbi.v3.core.Jdbi
import org.jdbi.v3.core.config.Configurable
import org.jdbi.v3.core.config.JdbiConfig
import org.jdbi.v3.core.extension.ExtensionCallback
import org.jdbi.v3.core.extension.ExtensionConsumer
import org.jdbi.v3.core.kotlin.internal.KotlinPropertyArguments
import org.jdbi.v3.core.qualifier.Qualifier
import org.jdbi.v3.core.result.ResultBearing
import org.jdbi.v3.core.result.ResultIterable
import org.jdbi.v3.core.statement.SqlStatement
import org.jdbi.v3.core.statement.StatementContext
import org.jdbi.v3.meta.Alpha
import org.jdbi.v3.meta.Beta
import java.util.function.Consumer
import kotlin.coroutines.CoroutineContext
import kotlin.reflect.KAnnotatedElement
import kotlin.reflect.KClass
import kotlin.reflect.full.findAnnotation

private const val METADATA_FQ_NAME = "kotlin.Metadata"

/**
 * Returns true if the [Class][java.lang.Class] object represents a Kotlin class.
 *
 * @return True if this is a Kotlin class.
 */
fun Class<*>.isKotlinClass() = this.annotations.singleOrNull { it.annotationClass.java.name == METADATA_FQ_NAME } != null

internal fun KClass<*>.simpleName(): String = this.simpleName ?: this::java.javaClass.simpleName

/**
 * Convenience helper to use any [KClass] for config lookup.
 *
 * @see StatementContext.getConfig
 */
fun <T : JdbiConfig<T>> StatementContext.getConfig(kClass: KClass<T>): T = this.getConfig(kClass.java)

/**
 * Use a reified parameter to map the result.
 *
 * @param T The type to use for mapping.
 */
inline fun <reified T : Any> ResultBearing.mapTo(): ResultIterable<T> = this.mapTo(T::class)

/**
 * Map to a Kotlin class.
 *
 * @param kClass the type to map the result set rows to.
 *
 * @see ResultBearing.mapTo
 */
fun <T : Any> ResultBearing.mapTo(kClass: KClass<T>): ResultIterable<T> = this.mapTo(kClass.java)

/**
 * Stream all the rows of the result set out with a [Sequence]. Handles closing of the underlying iterator.
 *
 * ```
 * handle.createQuery(...).mapTo<Result>().useSequence { var firstResult = it.first() }
 * ```
 */
inline fun <O : Any> ResultIterable<O>.useSequence(block: (Sequence<O>) -> Unit): Unit = this.iterator().use { block(it.asSequence()) }

/**
 * Bind all the member properties of a given Kotlin object.
 *
 * @param prefix A prefix for the property names.
 * @param obj The object to bind.
 */
@Beta
fun <This : SqlStatement<This>> SqlStatement<This>.bindKotlin(prefix: String, obj: Any): This =
    this.bindNamedArgumentFinder(KotlinPropertyArguments(obj, prefix))

/**
 * Bind all the member properties of a given Kotlin object.
 *
 * @param obj The object to bind.
 */
@Beta
fun <This : SqlStatement<This>> SqlStatement<This>.bindKotlin(obj: Any): This = this.bindNamedArgumentFinder(KotlinPropertyArguments(obj))

/**
 * Convenience method for [Configurable.configure] using Kotlin class syntax.
 *
 * @param configClass – the configuration type
 * @param configurer – consumer that will be passed the configuration object
 * @see Configurable.configure
 */
@Beta
fun <This : Configurable<This>, C : JdbiConfig<C>> Configurable<This>.configure(configClass: KClass<C>, configurer: Consumer<C>): This =
    this.configure(configClass.java, configurer)

/**
 * A convenience method which opens an extension of the given type, yields it to a callback, and returns the result
 * of the callback. A handle is opened if needed by the extension, and closed before returning to the caller.
 *
 * @param extensionType the type of extension.
 * @param callback      a callback which will receive the extension.
 * @param <R> the return type
 * @param <E> the extension type
 * @param <X> the exception type optionally thrown by the callback
 * @return the value returned by the callback.
 * @throws org.jdbi.v3.core.extension.NoSuchExtensionException if no [org.jdbi.v3.core.extension.ExtensionFactory]
 * is registered which supports the given extension type.
 * @throws X                        if thrown by the callback.
 */
fun <E : Any, R, X : Exception> Jdbi.withExtension(extensionType: KClass<E>, callback: ExtensionCallback<R, E, X>): R =
    withExtension(extensionType.java, callback)

/**
 * A convenience method which opens an extension of the given type, and yields it to a callback. A handle is opened
 * if needed by the extention, and closed before returning to the caller.
 *
 * @param extensionType the type of extension
 * @param callback      a callback which will receive the extension
 * @param <E>           the extension type
 * @param <X>           the exception type optionally thrown by the callback
 * @throws org.jdbi.v3.core.extension.NoSuchExtensionException if no [org.jdbi.v3.core.extension.ExtensionFactory]
 * is registered which supports the given extension type.
 * @throws X                        if thrown by the callback.
 */
fun <E : Any, X : Exception> Jdbi.useExtension(extensionType: KClass<E>, callback: ExtensionConsumer<E, X>): Unit = useExtension(extensionType.java, callback)

/**
 * Returns the set of qualifying annotations on the given Kotlin elements.
 * @param elements the annotated element. Null elements are ignored.
 * @return the set of qualifying annotations on the given elements.
 */
fun getQualifiers(vararg elements: KAnnotatedElement?): Set<Annotation> {
    return elements.filterNotNull()
        .flatMap { element -> element.annotations }
        .filter { anno -> anno.annotationClass.findAnnotation<Qualifier>() != null }
        .toSet()
}

/**
 * Returns a [CoroutineContext.Element] instance representing this Jdbi object which can be used to manage
 * handles from this Jdbi in a coroutine context.
 *
 * @return A [CoroutineContext.Element] object. No attempt should be made to inspect or use this object directly. Coroutines
 * that share a context which contains this element will share any derived object such as [org.jdbi.v3.core.Handle] or
 * extension objects.
 *
 * @since 3.42.0
 */
@Alpha
fun Jdbi.inCoroutineContext(handle: Handle? = null): CoroutineContext.Element {
    val handleScope = this.handleScope
    if (handleScope is CoroutineHandleScope) {
        handleScope.handle = handle
        return handleScope
    } else {
        throw IllegalStateException("Coroutine support was not enabled!")
    }
}
